Implemented support for parsing multiple exports

Aaron-Pool opened the pull request:
## Related issues / Problem

https://github.com/vue-styleguidist/vue-styleguidist/pull/665

## What I did

Updated `vue-docgen-api` to `4.1.1`, then converted loader to use `parseMulti` to support multiple exports in a file. Added a test for multiple exports as well.

Also, two of your other tests were operating under the assumption that default exports weren't supported by vue-docgen-api, but I _think_ that's not the case. It seems like it just doesn't support `export default` when used with `Vue.extends({})` (which should maybe be a case @elevatebart should add support for, since I think that's a common method to get correct typescript intellisense?). I fixed those tests and also added cases for mixed exports (default and named). Those changes might be outside of the scope of this PR, so I put them in a separate commit and can move them out, if requested.

## Checklist


- [x] Added one or more test
- [x] This PR does not includes any other changes that is unrelated to the problem/feature
  ~~- Again, see note about test changes above, I can move that commit out if need be.~~ (I just removed this commit for now, to reduce the noise)

@pocka @Shilman @elevateBart

Aaron-Pool committed:
feat: added support for multiple exports

Aaron-Pool committed:
test: added further test cases for named and default exports

shilman commented:
@Aaron-Pool AWESOME!! Thanks so much for putting this together!

One comment is that the peer dep for `vue-docgen-loader` should be updated to "^4.1" and it should probably be a major version bump for `vue-docgen-loader`.

@pocka Would love to get this merged ASAP 🙏🙏🙏 😁

Aaron-Pool commented:
@shilman Ah, good point on the version number. I'll get that fixed. 

For the second point though, why a major version bump? It shouldn't break anyone current usage--unless I misunderstand something. It's only adding additional docgen info to exports that were previously getting none. I thought major version bumps were reserved for incompatible API changes?

pocka commented on test/fixtures/basicDefault.vue.js:
```diff
@@ -0,0 +1,14 @@
+import Vue from 'vue'
+
+export default {
```
@Aaron-Pool 
Though not tested, I think this would be turned to `export default {/* ... */}; component.__docgenInfo = // ...` and it'll throw an error, because the loader treat `exportName: 'default'` as SFCs. And if we could fix that, the loader can't handle this case because there is no way to attach `__docgenInfo` to the default export object... 😞 

Adding a warning about the default export on README then create an issue is the safest way, IMO. What do you think?

Aaron-Pool replied:
Hmm, I think my changes have have tests passing on both SFC and non-SFC cases? So I'm confused on exactly what you're saying 🤔

Aaron-Pool replied:
Oh wait, I think I see what you're saying. Let me dig in a bit more.

pocka replied:
Sorry for my poor word choices... I mean "`vue-docgen-api` will parse the code successfully, but the loader would append incorrect code, and this will cause a run-time error."

JavaScript code after the process will be looked like this:

```js
import Vue from 'vue'

export default {
  props: {
    /**
     * Foo the number.
     */
    foo: {
      type: Number,
      default: -1
    }
  },
  template: '<span>{{foo}}</span>'
}
// this object (docgen info) is valid, but we try to add a property to a variable not exist
;component.__docgenInfo = {/* ... */}```

Aaron-Pool replied:
Right, I see that now. Is there a reason we couldn't do a simple regex replace of
```js
export default {
    ...definition
}
```
with 
```js
const component = {
    ...definition
}
export default component;
```
Since the name of the default export, by definition of the esmodule spec, doesn't matter?

pocka replied:
>  Is there a reason we couldn't do a simple regex replace 

No, we can replace `export default <expression>` (I don't remember why I haven't implemented that transformation.) Now I think this is an existing bug and will fix this in another branch.

Aaron-Pool replied:
@pocka I wrote a simple jscodeshift transform before I saw your response. It occurred to me that it might be more reliable than a regex, in case someone uses a weird mangled or minified source code. If you'd like to use it, [you're welcome to it](https://astexplorer.net/#/gist/2dfbbbc37e0f53d31bd6f2a1bedca9c0/latest), but if you have other ideas I understand that 👌

pocka replied:
Great thanks!

pocka commented:
@shilman 
> One comment is that the peer dep for vue-docgen-loader should be updated to "^4.1" and it should probably be a major version bump for vue-docgen-loader.

Oh, my bad:scream: It should be `>=3` to support 3.x and 4.x~.

shilman commented:
@Aaron-Pool major version bump since `vue-docgen-api` is a peer dep, not a direct dependency, so upgrading `vue-docgen-loader` also requires upgrading `vue-docgen-api` manually. I'm actually not sure the semver rules about this, but it wouldn't hurt to do a major version bump.

Certainly the current code here is not compatible with 3.x because `parseMulti` didn't exist in 3.x...

pocka commented on src/index.js:
```diff
@@ -28,21 +28,24 @@ module.exports = async function(content, map) {
     const infoOrPromise = isSFC
       ? docgen.parse(this.resourcePath, options.docgenOptions)
-      : docgen.parseSource(content, this.resourcePath, options.docgenOptions)
+      : docgen.parseMulti(this.resourcePath, options.docgenOptions)
```
@Aaron-Pool 
As [@shilman commented](https://github.com/pocka/vue-docgen-loader/pull/4#issuecomment-562456067), this line would cause an error because `parseMulti` does not exist in `vue-docgen-api@3.x`. Please add an existence check and fallback for `parseSource` (or `parse`).

Aaron-Pool replied:
Right, sorry, I've done very little library development, so I'm not super familiar with how peer dependencies work. I can definitely do what you just described though 👍

Aaron-Pool committed:
Revert "test: added further test cases for named and default exports"

This reverts commit 5211a164d84803743ee8a806ea55f3ed2e1cde07.

Aaron-Pool committed:
fix: corrected docgen-api verison and added parseMulti check

Aaron-Pool committed:
chore: readded erroneously removed peer dependency

Aaron-Pool commented:
Ok, I went ahead and removed the extra tests about export default just to reduce the noise on the PR. I also updated the peer dependency as requested.

pocka commented:
Published in v1.2.0

For the `export default` problem, I opened https://github.com/pocka/vue-docgen-loader/issues/5

Aaron-Pool commented:
@pocka Thanks for the merge. And I honestly have no idea how `vue-docgen-api` somehow ended up in the regular dependency list, sorry about that 🤔 Also apologies for the linter errors, I was assuming it would reject any commits that had linting issues. Kind of a messy PR on my part, thanks for picking up the slack 🙌

pocka commented:
@Aaron-Pool 
No, problem! and Thank you so much for creating PR :blush: (and using PR template!)
